### DISCUSSION TITLE
Improvement ideas for rest API

### DISCUSSION THREAD

**Discussion by JuuZoop** [jmstoolbox/jmstoolbox#141] [Ideas]
@titou10titou10 This seems like a very promising Open source program for MQ interactions. I especially think the rest API is great for integration testing purposes.

There are couple of things that could be added towards REST API
- Message selector support for message browsing
- Message information in POST message response payload

**Message selector support for message browsing**

Currently it is only possible to select all messages within a queue. You can only pass a query parameter to limit amount of retrieved messages. When there is a test case that requires obtaining a specific message it would be very beneficial if the REST API offered a way to pass a message selector to the JMSToolBox. With message selector it would be possible to filter messages that are retrieved.

**Message information in POST message response payload**

Request:

POST http://localhost:9998/rest/message/{QueueManager}/{Queue}
 {
  "deliveryMode" : "PERSISTENT",
  "priority" : "1",
  "deliveryDelay" : "",
  "timeToLive" : "",
  "jmsCorrelationID" : "9af59e05-4c2f-4dd6-b358-38fad6402074",
  "jmsType" : "SomeType",
  "type" : "TEXT",
  "replyTo" : "",
  "payloadText" : "Some example text"
 }

This will yield a response:

HTTP/1.1 201 Created
Date: Thu, 11 Aug 2022 20:08:21 GMT
Content-Length: 0
Server: Jetty(10.0.6)

The response itself does not contain other information than the MQ sending was a success or a failure. It would be good to also have updated message information and message header information in the response payload. This way client would be able to obtain for example a JMSMessageID that could be used for test validation purposes. These updated message information set by the JMS Provider should be able to get from Message object after successful message sending.

**Comment by titou10titou10** [comment]
Adding message selector should not be a problem
[UPDATE]
The difficulty I see is the selector string may be quite "long" and can contain any of the UTF-8 character (obviously "equal", "quotes", "commas"...). It may not be a good idea to use such strings as query parameters in an HTTP GET request. 
It seems a better idea to add this feature as an HTTP POST request (instead of a GET) and pass the selector in the body payload of the request where there is no length nor character content constraints...

As for returning data from the message post, JTB performs a`"MessageProducer.send()"`call to post the message to the queue from the JMS API. This method does not return any data
In order to retrieve some meta data from the just created message (ege JMSMessageID..), JTB should "get/read" it in some way. 
JTB does not have enough information to retrieve the exact message it just send, and worse, the message could not appear in the queue just after JTB post it. This is up to the QM and, of course, the`deliveryDelay`and`timeToLive`JMS properties could interfere...

**Reply by JuuZoop** [reply]
In my understanding an underlying JMSMessage object javax.jms.Message would update after successful sending and after sending it could be queried for messageinformation such as JMSMessageId. 

I checked the JMSToolBox source code and noticed that REST service operation /rest/message/{sessionName}/{destinationName}/{templateName} would return information about the message that is sent based on template. This output seems to be JSON format of MessageOutput class object that is created from JTBMessage class object. I would have checked whether I could get a JMSMessageId to appear in response body when sending a message by utlizing a message sending through a template from REST-API. However I seem only to get a "HTTP/1.1 500 Server Error" response even though I use the template name that I have created through JMSToolBox GUI.

I may try to inspect this further if I manage to get this project running within Eclipse.

**Reply by titou10titou10** [reply]
> In my understanding an underlying JMSMessage object javax.jms.Message would update after successful sending and after sending it could be queried for messageinformation such as JMSMessageId.

[UPDATE]
~~Humm. Could you point me in the JMS 2.0 specs where this is specified?~~
It seems you are right, I found this is the JMS 2.0a specs:

    3.4.3. JMSMessageID
        The JMSMessageID header field contains a value that uniquely identifies each message sent by a provider.
        When a message is sent, JMSMessageID is ignored. When the send method returns it contains a provider-assigned value.

I will have to give it a try (Same with JMSTimestamp, JMSExpiration...)

> I checked the JMSToolBox source code and noticed that REST service operation /rest/message/{sessionName}/{destinationName}/{templateName} would return information about the message that is sent based on template. This output seems to be JSON format of MessageOutput class object that is created from JTBMessage class object.

What is returned is the message send by JTB to the QM after JTB read the templates and have solved the template variables  not the message "re-read" from the QM. And of course, the message returned has no properties internally set by the QM like JMSMessageID

> I would have checked whether I could get a JMSMessageId to appear in response body when sending a message by utlizing a message sending through a template from REST-API. However I seem only to get a "HTTP/1.1 500 Server Error" response even though I use the template name that I have created through JMSToolBox GUI.

Possibly a bug, post JTB logs here

**Reply by JuuZoop** [reply]
I found a following in [Oracle JavaEE Specification regarding JMSMessageID method in javax.jms.message class:](https://docs.oracle.com/javaee/7/api/javax/jms/Message.html#getJMSMessageID--):
`When a message is sent, JMSMessageID can be ignored. When the send or publish method returns, it contains a provider-assigned value. `

Here are some log snippets:
2022-08-12 18:27:09:852 DEBUG [o.j.r.r.i18n                ] RESTEASY002315: PathInfo: /rest/message/SOME_QUEUE_MANAGER/SOME.TEST.QUEUE/testtemplate
2022-08-12 18:27:09:853 DEBUG [o.t.j.r.s.MessageServices   ] postMessageTemplate. sessionName=SOME_QUEUE_MANAGER destinationName=SOME.TEST.QUEUE templateName=testtemplate
2022-08-12 18:27:09:853 DEBUG [j.c.ExternalConnectorManager] postMessageTemplate
2022-08-12 18:27:09:853 DEBUG [o.t.j.j.m.JTBConnection     ] connect 'JTBConnection [jtbSessionClientType=REST, isConnected()=true]'
2022-08-12 18:27:09:853 DEBUG [o.t.j.t.TemplatesManager    ] getTemplateFromName 'testtemplate'

It seems that no exception is logged which is strange when comparing this with a source code.

**Comment by titou10titou10** [comment]
our posts crossed..lol

**Comment by titou10titou10** [comment]
For your error, nothing also in the`"{user}\.jtb\.metadata\.log"` file (Where the eclipse exception are tracked)?

**Reply by JuuZoop** [reply]
The `"{user}\.jtb\.metadata\.log"` seems to be only containing one log entry for today which is over an hour ago. It is something about workspace space exiting:
!ENTRY org.eclipse.core.resources 2 10035 2022-08-12 17:18:14.573
!MESSAGE The workspace will exit with unsaved changes in this session.
!SESSION 2022-08-12 17:18:56.182 -----------------------------------------------

The previous entry in that log is from yesterday.

**Comment by titou10titou10** [comment]
I've reproduced the HTTP 500 on my side. 
It's a core eclipse exception when JTB tries to locate the template on the file system in TemplateManager

https://github.com/jmstoolbox/jmstoolbox/blob/3c446f88e0fec8f8054c293f30f41d984a7409d2/org.titou10.jtb.core/src/org/titou10/jtb/template/TemplatesManager.java#L519

I guess the bug came from the newer version of eclipse RCP
I must find time to fix this first

**Comment by titou10titou10** [comment]
I've committed a fix for the HTTP 500 when posting a message from template

**Reply by JuuZoop** [reply]
Great! I can also test that on next week when I get the project set up.

**Comment by titou10titou10** [comment]
I've committed the change to receive all info from the freshly pushed message. Now the "Message" received in return looks like this:

    {
       "jtbMessageType": "TEXT",
       "jmsMessageID": "ID:414d51204d5120202020202020202020f9aef762026e0040",
       "jmsPriority": 4,
       "jmsDestination": "queue:///DEV.QUEUE.1",
       "jmsCorrelationID": "abcd",
       "jmsDeliveryMode": "PERSISTENT",
       "jmsDeliveryTime": "2022-08-13 10:43:14.954",
       "jmsTimestamp": "2022-08-13 10:43:14.954",
       "jmsRedelivered": false,
       "properties":    {
          "JMS_IBM_PutDate": "20220813",
          "JMSXDeliveryCount": "0",
          "JMSXUserID": "app         ",
          "JMS_IBM_PutTime": "14431474",
          "JMSXAppID": "JMSToolBox                  ",
         "JMS_IBM_PutApplType": "28"
       },
       "payloadText": "hello"
     }

I've also [updated the wiki](https://github.com/jmstoolbox/jmstoolbox/wiki/1.8-REST-Services)

**Reply by JuuZoop** [reply]
This looks awesome! Kinda like everything that could be required for integration testing purposes :)
Would it be good idea to also utilize this same data structure on GET and PUT operations that browse and get messages from the queue?

Overall this REST API is looking to be a way to go with different integration testing tools. 
Eg SoapUI has this deprecated support of interacting message queues with HermesJMS tool which has not been maintained in ages and I guess also has the source code not available for public. Now if you want to use SoapUI you could just utilize the basic REST request step and interact message queues with JMSToolBox REST API.

BTW I also notice that there are SoapUI plugin PoC conducted in JMSToolBox project. I have written a SoapUI plugin once and it was kinda tedious to get started with that due to lack of sufficient official documentation. I also had to rely on this so called "old style" plugin structure because the up to date plugin structure would require signing from authors or otherwise the plugin would not load. Good thing is that latest versions of SoapUI open source have ditched this signing requirement and new plugin structure plugins could now be loaded without restrictions.

**Reply by titou10titou10** [reply]
> Would it be good idea to also utilize this same data structure on GET and PUT operations that browse and get messages from the queue?

This is the case, the same "Message" structure is used for every message returned to the client

> Now if you want to use SoapUI you could just utilize the basic REST request step and interact message queues with JMSToolBox REST API.

I myself use SoapUI "standard" REST features for testing JTB  REST features. Code to integrate HermesJMS is very depply burrowed inside SoapUI code. At least it was the case 5 years ago

> BTW I also notice that there are SoapUI plugin PoC conducted in JMSToolBox project. 

Yes, I talked with the SoapUI guys years ago. I didn't continue because of the lack of interest of such plugin, the signing process that was a pain (even the bypass for testing was eurk) amd the complexity of writing such plugin. 
Good to know things have improved compared to 5 years ago
I'm not sure if there is a real value added and interest for such plugin. Pull Requests are welcome lol

**Comment by titou10titou10** [comment]
I've publishes v5.15.0 with the 2 new features

**Reply by JuuZoop** [reply]
I tested the new release. Looks good.
- Message selector is working. I am able to filter messages
- The new model for message is consistend among different resources/methods and I am also able to get all relevant information when sending a new message
- Template based message sending now works also. The <templatename>.jtb just has to be outside of the Templates folder in order it to be detected.
